### PR TITLE
Hamiltonian Monte Carlo

### PR DESCRIPTION
**Pull request by joezuntz**
I've written up my Hamiltonian Monte Carlo experiment, and included the code and chains (sorry for the large diff the latter yields).

- Code: test notebook, hmc module, executable, cov estimate
- Results: chain files, constraint plot, plotter notebook
- Paper: section in paper, bib file
- Misc: Makefile, gitignore

### CODE REVIEW

**Comment by EiffL** [comment]
Awesome :-D So... if I understand correctly you are saying you are beating Cobaya by a factor 2 with vanilla HMC :-) Now we just need to do some timings of logp and grad on different architectures, and that will give us a lower bound on expected speed up.

I see you are running the HMC sampler in parallel, what configuration did you use? something like a process per core, or a process per node? and how many ranks?

**Comment by joezuntz** [comment]
> if I understand correctly you are saying you are beating Cobaya by a factor 2 with vanilla HMC

I was careful not to say that :-)  HMC took several times longer on this system, in total, and you can see the HMC is noisier.  I think this marks that this is a productive way to go, especially as we move to higher dimensionality where it should scale really really well, but this wasn't supposed to be a full comparison.

I'll add details on the run to the paper - for both samplers I used 2 processes with 4 threads/process. This was on a local cluster.

File: notebooks/desy1.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    # returns -posterior P and -dL/dP
+    def posterior_and_gradient(self, p):
+        # theory C_ell prediction
+        cl = theory_mean(p, *self.args)
+        # d C_ell / d p
+        j = theory_jacobian(p, *self.args).T
+        d = cl - self.data_mean
+        dlogL_dCl = -self.data_inv_cov @ d
+        logL = 0.5 * d @ dlogL_dCl
+        dlogL_dp = j @ dlogL_dCl
```
Comment (EiffL) [inline notebooks/desy1.py:218]: ah yeah, here we should be able to get some extra speed up, I *suspect* doing the jacobian computation separately here is going to be slower than the optimized `jax.value_and_grad` :
https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Evaluate-a-function-and-its-gradient-using-value_and_grad

**Review by EiffL — APPROVED** [review]
This is awesome! only made one suggestion for potential speed up, but then again, this would only affect speed, not number of samples.

**Comment by EiffL** [comment]
ok ok :-) So, am I understanding correctly that you have 2 HMC chains running in parallel then? This is quite good :-) because I have good hopes of being able to run like 128 chains at once on a decent size GPU.

**Comment by joezuntz** [comment]
Yep, exactly, and that way I could do the Gelman-Rubin test between them to check for convergence.  

Is it more efficient to evaluate lots of likelihoods at once or to use lots of GPU threads for one likelihood?

It looks like there are at least some algorithms for ensembles of derivative-aware samples:
https://link.springer.com/article/10.1007/s11222-017-9730-1

**Comment by EiffL** [comment]
> Is it more efficient to evaluate lots of likelihoods at once or to use lots of GPU threads for one likelihood?

The former is the answer. Because for one computation, you can't really use more threads to process it. But you can vectorize/batch the computation of a bunch of likelihoods, without loosing speed. It takes the same time to compute 2 likelihoods in parralel than 1.  And this is true up until you reach the memory limit of the GPU. 

So, in the future I imagine we will be using things like cori-gpu nodes, with 4 large 32GB GPUs on each node. Each GPU could definitly fit 512 DES Y1 likelihood evals in parallel, so that one cori-gpu node would give you of the order of 2048 parallel chains.  And for example one DES Y1 likelihood probably takes under 100ms to evaluate. So that's about 20,000 likelihood evals per seconds on 1 node, give or take.

**Comment by joezuntz** [comment]
Forgot about this!  Are you happy for me to merge?

**Comment by EiffL** [comment]
oh yes yes. do click the green button if you have enough rights, let me know otherwise :-)

**Comment by joezuntz** [comment]
Green button clicked!
